[no ticket][risk=no] Update pull request template (#7586)
all-of-us/workbench: .github/PULL_REQUEST_TEMPLATE.md

Description:
<!--
Replace this template with your PR description.
Please remember to keep in mind the security levels outlined in
[CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md) and to
include a risk tag of the form `[risk=no|low|moderate|severe]` in the PR title

* **no**: None
* **low**: Low chance of potential impact to, or exposure of patient data
* **moderate**: Moderate chance of potential impact to, or exposure of patient data

> * **severe**: Severe chance of potential impact to, or exposure of patient data

Author: Moved into a succinct checklist item.

Please also:

> * Get thumbs from reviewer(s)

Author: This is a hard requirement to merge, so it doesn't need to be here.

> * Verify all tests go green, including CI tests

Author: Opinion: All items in the PR checklist should be actionable at the time of creating the PR.

Reviewer: I would prefer a reminder for this instead of a checklist item:
> If you decide to merge with any failing checks, add an explanatory comment before doing so.

This says, "We are going to allow you to make this judgment call, but need an explanation both for trust building and compliance."

Author: Fair. I'll add a reminder.

> - [ ] If this PR is associated with a Jira issue, I have included an issue number tag in the PR title as outlined in [CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md)

Author: This is new.

Reviewer: ```suggestion
- [ ] I have included an issue ID or "no ticket" in the PR title as outlined in [CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md)
```
Suggesting "no ticket" may help us be consistent in case we ever want to search for these.

Author: Good idea, done

> - [ ] This PR includes appropriate automated tests, and I have documented any behavior that cannot be tested with code

Author: - Adds a basic QA step to replace "I have run and tested this change locally, and my testing process is described here"
- Changes `unit tests` to `automated tests`
- Call out anything that cannot be tested

> - [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

Author: Note: It would be nice if, someday, this were an automated PR check for feature flag modification that reminded us to do this.

-->

**Description**

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have added explanatory comments where the logic is not obvious
- [ ] I have run and tested this change locally, and my testing process is described here
- [ ] If this PR is associated with a Jira issue, I have included an issue number tag in the PR title as outlined in [CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md)
- [ ] I have included a risk tag of the form `[risk=no|low|moderate|severe]` in the PR title as outlined in [CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md)
- [ ] I have manually tested this change
- [ ] This PR includes appropriate automated tests, and I have documented any behavior that cannot be tested with code
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

> - [ ] If this includes an API change, I have run the E2E tests on this change against my local server with [yarn test-local](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples) because this PR won't be covered by the CircleCI tests 

Author: > this PR won't be covered by the CircleCI tests

I don't think this is true, so I've removed this item.

Reviewer 2: Maybe the wording could be improved?

This is intended to convey that the e2e tests that are run pre-merge use the PR's UI but the Test env's API.  So it is true that the PR's API changes won't get tested there.

The main-branch post-merge e2e tests do cover the newest API code.

Author: Wow, I thought that we were re-releasing the test API on every PR run. This helps explain why some devs use the test env as a development env.

Why do we even run e2e tests on PR checks if they only check half the code? In some cases they're guaranteed to fail.

I'd like to add an item that describes what the developer should do here. "Run all the tests" would be reasonable if our tests were fast and consistent, but since they aren't that seems inconvenient. "Run all relevant tests" seems like a compromise but requires developers to be loosely aware of all our E2E tests.

What do you do each PR?

Reviewer 2: > Why do we even run e2e tests on PR checks if they only check half the code?
That's a valid way of looking at it, but I think in practice *most* of the tests are relevant *most* of the time.

I like "run all relevant tests".  It's true that this requires a greater awareness of which e2e tests might affect API code, but I think that's a good thing - at least in our current framework.

If I make an API change in a PR, I always also run it locally to test that it's doing what I expect.  I don't necessarily run local e2e tests in the way we're talking about - but maybe I should.

Reviewer 2: > some devs use the test env as a development env.
I would like to explore this some more (not here) because I'm not sure I understand exactly what difference you're describing

Author: Re-added with updated wording 👍

- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P)
- [ ] If this change impacts deployment safety (e.g. removing/altering APIs which are in use) I have documented these in the description
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and updated the appropriate API consumers
  * AoU UI
  * [Perf tests](https://github.com/broadinstitute/mcnulty/blob/develop/src/test/scala/services/AoU.scala)
  * [Researcher Directory export](https://github.com/all-of-us/workbench/wiki/Researcher-Directory-(RDR-export))
  * Cron tasks - for Offline*Controllers

>   * SumoLogic - for EgressAlert 

Author: Is this something that happens often enough to warrant checking every single PR?

This seems like a strange place to document our API consumers. Most of the breakages this item is trying to detect would be picked up by our type system or our tests. We rarely modify SumoLogic.

Reviewer 2: I remember that we did think this was worthwhile when adding it.  May be worth revisiting now, of course.

Perf tests -> no longer exist, so we can remove that.
RDR Export is a cron task so we can group with them -> we can catch problems here by looking at cron error logs.  (we should also be more diligent about this)

OK, seems we can probably remove this checkbox.

But I would like to keep the deployment-safety check.

Author: I re-added the deployment-safety check.

Description:
<!--
Replace this template with your PR description.
Please remember to keep in mind the security levels outlined in
[CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md) and to
include a risk tag of the form `[risk=no|low|moderate|severe]` in the PR title

* **no**: None
* **low**: Low chance of potential impact to, or exposure of patient data
* **moderate**: Moderate chance of potential impact to, or exposure of patient data
* **severe**: Severe chance of potential impact to, or exposure of patient data

Please also:

* Get thumbs from reviewer(s)
* Verify all tests go green, including CI tests
-->

**Description**

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests

> - [ ] I have added explanatory comments where the logic is not obvious

Author: True, this covers explanatory PR comments as well. I'll leave it in.

> - [ ] I have manually tested this change

Author: I've only been doing this for "non-obvious" PRs but "obvious" is very subjective and it's easy enough to post a sentence explaining how to verify a PR. I'll re-add this.

[Relevant below comment](https://github.com/all-of-us/workbench/pull/7586#discussion_r1168816416).

> - [ ] If this addresses a bug, I have described the steps to reproduce the issue above

Author: From @NehaBroad's comment above:

> sometimes we can get bugs that are tricky to reproduce or we find some edge cases in our code and open a quick PR to fix that

I think this refers to `no-ticket` PRs that address small bugs.

However, re-adding `I have manually tested this change and my testing process is described above` would cover this since you need to describe the steps to manually verify the fix. I'm going to swap the two.

Description:
<!--
Replace this template with your PR description.
Please remember to keep in mind the security levels outlined in
[CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md) and to
include a risk tag of the form `[risk=no|low|moderate|severe]` in the PR title

* **no**: None
* **low**: Low chance of potential impact to, or exposure of patient data
* **moderate**: Moderate chance of potential impact to, or exposure of patient data
* **severe**: Severe chance of potential impact to, or exposure of patient data

Please also:
**Description**

* Get thumbs from reviewer(s)
* Verify all tests go green, including CI tests
<!--
Reminder: If you decide to merge with any failing checks, add an explanatory comment before doing so.
-->

---
**PR checklist**

> - [ ] This PR meets the Acceptance Criteria in the JIRA story

Reviewer 2: The intention here is to encourage an extra check of the AC before closing a story, to address a habit of doing **most** of a story, but missing some pieces.  TBH I continue to feel that's an area where we need to improve.

But this is not relevant to all PRs, so I'm not sure that it belongs here.

How about something like this: "if this PR is intended to complete a JIRA story, I have checked that all AC are met for that story"
